### DISCUSSION TITLE
Setup encrypted userinfo response

### DISCUSSION THREAD

**Discussion by ricklambrechts** [keycloak/keycloak#27663] [Q&A]
Encrypted Userinfo response is added in https://github.com/keycloak/keycloak/pull/10519.

Now I am wondering where I can set the public key for client. The userinfo needs to be encrypted with this public key so that the client application can decrypt it.

The client specific 'Fine grain OpenID Connect configuration' contains the needed settings for the algorithms.
![image](https://github.com/keycloak/keycloak/assets/1367665/2b003aaf-9b3d-4689-9107-7ebc465c7fc6)

But I am missing the place where I can set the public key per client. Does someone has setup ecrypted userinfo or knows where the setting is located?

As expected the log is showing:

```
2024-03-07 13:00:30,507 WARN  [org.keycloak.keys.loader.ClientPublicKeyLoader] (executor-thread-227) Unable to retrieve publicKey of client 'test-zsm-client' for the specified purpose other than verifying signature
2024-03-07 13:00:30,508 WARN  [org.keycloak.keys.infinispan.InfinispanPublicKeyStorageProvider] (executor-thread-227) PublicKey wasn't found in the storage. Requested kid: 'null' . Available kids: '[]'
```

**Comment by justin-tay** [comment]
This should be configured under `Keys` in `Client details`.

Typically your client will expose the required keys to Keycloak by hosting a JWKS endpoint.

![keys](https://github.com/keycloak/keycloak/assets/49700559/413e2f8c-30f2-4fe8-951d-07fecc6faba8)

**Comment by ricklambrechts** [comment]
Indeed. Also setting log level to debug mode shows more relevant logging.

If someone also has a problem with it:
- kid should be: 
  - "kid": null,
- use should be:
  - "use": "enc"'


With debug logging you see the request being made to the JWKS url and if it is successfull or not.
And I expected that the JWKS keys gonna be visible on the Keys page, but they are not. So in de administration panel you cannot see if the JWKS is loaded successfull.
